src/glarea.rs

```diff
@@ -0,0 +1,208 @@
+//! # GLArea Sample
+//!
+//! This sample demonstrates how to use GLAreas and OpenGL
+
+extern crate gtk;
+extern crate libc;
+
+// make moving clones into closures more convenient
+macro_rules! clone {
+    ($($n:ident),+; || $body:block) => (
+        {
+            $( let $n = $n.clone(); )+
+            move || { $body }
+        }
+    );
+    ($($n:ident),+; |$($p:ident),+| $body:block) => (
+        {
+            $( let $n = $n.clone(); )+
+            move |$($p),+| { $body }
+        }
+    );
+}
+
+#[cfg(feature = "opengl")]
+extern crate epoxy;
+
+// Based off of the static-triangle example from gl-rs
+#[cfg(feature = "opengl")]
+mod example {
+    use std::mem;
+    use std::ptr;
+    use std::str;
+
+    use gtk;
+    use gtk::traits::*;
+    use gtk::signal::Inhibit;
+    use gtk::{GLArea, Window};
+
+    use epoxy;
+    use epoxy::types::*;
+    use epoxy::Gl;
+
+    fn compile_shader(src: &str, ty: GLenum) -> GLuint {
+        unsafe {
+            let shader = Gl.CreateShader(ty);
+            // Attempt to compile the shader
+            let psrc = src.as_ptr() as *const GLchar;
+            let len = src.len() as GLint;
+            Gl.ShaderSource(shader, 1, &psrc, &len);
+            Gl.CompileShader(shader);
+
+            // Get the compile status
+            let mut status = epoxy::FALSE as GLint;
+            Gl.GetShaderiv(shader, epoxy::COMPILE_STATUS, &mut status);
+
+            // Fail on error
+            if status != (epoxy::TRUE as GLint) {
+                let mut len = 0;
+                Gl.GetShaderiv(shader, epoxy::INFO_LOG_LENGTH, &mut len);
+                let mut buf = Vec::with_capacity(len as usize);
+                buf.set_len((len as usize) - 1); // Skip trailing null
+                Gl.GetShaderInfoLog(shader, len, ptr::null_mut(), buf.as_mut_ptr() as *mut GLchar);
+                panic!("Error compiling shader: {}",
+                       str::from_utf8(&buf).ok().expect("ShaderInfoLog not valid utf8"));
+            }
+
+            shader
+        }
+    }
+
+    fn link_program(vs: GLuint, fs: GLuint) -> GLuint {
+        unsafe {
+            let program = Gl.CreateProgram();
+            Gl.AttachShader(program, vs);
+            Gl.AttachShader(program, fs);
+            Gl.LinkProgram(program);
+
+            // Get the link status
+            let mut status = epoxy::FALSE as GLint;
+            Gl.GetProgramiv(program, epoxy::LINK_STATUS, &mut status);
+
+            // Fail on error
+            if status != (epoxy::TRUE as GLint) {
+                let mut len: GLint = 0;
+                Gl.GetProgramiv(program, epoxy::INFO_LOG_LENGTH, &mut len);
+                let mut buf = Vec::with_capacity(len as usize);
+                buf.set_len((len as usize) - 1); // Skip trailing null
```

Comment:
Okay that sounds good, will make that change. In general writing to a Vec within its allocated capacity should be "okay" though (for some definition of okay)?